cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +131,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
+	switch {
+	case regionSpecified && !resetSpecified:
+		// It doesn't make sense to specify a region unless we are resetting
+		// values here.
+		return errors.New("specifying a region when retrieving defaults for a setting is invalid")
+	case !regionSpecified && resetSpecified:
+		// It makes no sense to supply a positional arg that isn't a region if
+		// we are resetting a region, so we must have gotten an invalid region.
+		return errors.Errorf("invalid region specified: %q", arg)
+	case !regionSpecified && !resetSpecified:
+		// Phew, we can retrieve a value.
+		c.key = arg
+		c.action = c.getDefaults
+		return nil
+	case regionSpecified && resetSpecified:
+		// If a region was specified and reset was specified, we shouldn't have
+		// an extra arg. If it had an "=" in it, we should have handled it
+		// already.
+		return errors.New("cannot retrieve defaults for a key and reset args at the same time")
+	default:
+		// This should be unreachable.
+		return errors.New("unexpected input, please file a bug")
+	}
+}
+
+// handleExtraArgs handles the case where too many args were supplied.
+func (c *defaultsCommand) handleExtraArgs(args []string) error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+	numArgs := len(args)
+
+	// if we have a key=value pair here then something is wrong because the
+	// last positional arg is not one. We assume the user intended to get a
+	// value after setting them.
+	for _, arg := range args {
+		if strings.Contains(arg, "=") {
+			return errors.New("cannot set and retrieve default values simultaneously")
+		}
+	}
+
+	switch {
+	case numArgs == 2 && !regionSpecified && resetSpecified:
+		// It makes no sense to supply a positional arg that isn't a region if
+		// we are resetting a region, so we must have gotten an invalid region.
+		return errors.Errorf("invalid region specified: %q", args[0])
+	case !regionSpecified && !resetSpecified:
+		// If we land here it is because there are extraneous positional args.
+		return errors.New("can only retrieve defaults for one key or all")
+	default:
+		// Anything else just has extra positional args, which is invalid in an
+		// unreliably guessable way.
+		return errors.New("invalid input")
+	}
 }
 
 // Run implements part of the cmd.Command interface.
 func (c *defaultsCommand) Run(ctx *cmd.Context) error {
-	client, err := c.getAPI()
+	root, err := c.newAPIRoot()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	client := c.newDefaultsAPI(root)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	defer client.Close()
 
+	if len(c.resetKeys) > 0 {
+		err := c.resetDefaults(client, ctx)
+		if err != nil {
+			// We return this error naked as it is almost certainly going to be
+			// cmd.ErrSilent and the cmd.Command framework expects that back
+			// from cmd.Run if the process is blocked.
+			return err
```

Comment:
This is here because we used to run one action (cmd.Run) and return the call which in turn returned an error. With this collapse we might run more than one action (reset and set). Everywhere that things aren't collapsed expect to handle a blocked error as cmd.ErrSilent. So we want to return that rather than errors.Trace(err) if the process is blocked otherwise we don't error silently. It is really just playing along with the way things already work, but packing two commands into one. If it isn't cmd.ErrSilent the cmd.Command framework handles the error, IIUC.